### PR TITLE
Feature/replace flat file with nudb

### PR DESCRIPTION
**Pull request by Warchant**
### Description of the Change

- [x] remove flat_file implementation
- [x] preserve flat_file.hpp interface in block_storage.hpp
- [x] make block_storage.hpp pure virtual class
- [x] split block_query_test onto two tests:  
       - unit test: uses mock for block  storage
       - integration test: uses NuDB underneath
- [x] implement block_storage.hpp interface with NuDB
- [x] remove `last_id` from block_storage.hpp, because logically it is invalid. Instead, add `total_keys` method.
- [x] update `hyperledger/iroha-docker-develop` to `v2` - it has new CMake (3.10.2), more boost libraries (thread, chrono, program_options), integrated NuDB
- [x] fix all tests, few minor bugs and miss spellings

### Benefits

- new cmake version in develop image
- new implementation of Block Storage with use of NuDB
- now Block Storage stores blocks properly, from index #0

### Possible Drawbacks 

- NuDB is bad option to store blocks and must be considered as short-term solution.
- BlockStorage interface must be reviewed and redesigned with use of cryptography (to have a possibility to implement light clients)
- BlockStorage now has no block validation logic. I.e., when you open database first time, you need to check that all stored blocks are really blocks, have valid signatures, and re-build WSV (or check its consistency, TBD) 

### Usage Examples or Tests *[optional]*

- [./test/module/vendor/nudb_usage.cpp](./test/module/vendor/nudb_usage.cpp)

### Alternate Designs *[optional]*

https://github.com/hyperledger/iroha/issues/774

### CODE REVIEW

**Comment by bakhtin** [comment]
SonarQube analysis reported 48 issues
* ![MINOR][MINOR] 14 minor
* ![INFO][INFO] 34 info

#### Top 10 issues

1. ![MINOR][MINOR] [block_storage_nudb.cpp#L115](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/irohad/ametsuchi/impl/block_storage_nudb.cpp#L115): Define a constant instead of duplicating this literal false 3 times. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3AStringLiteralDuplicated)
1. ![MINOR][MINOR] [redis_block_query.hpp#L63](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/irohad/ametsuchi/impl/redis_block_query.hpp#L63): Unused private function: 'RedisBlockQuery::getBlockIds' [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedPrivateFunction)
1. ![MINOR][MINOR] [redis_block_query.hpp#L71](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/irohad/ametsuchi/impl/redis_block_query.hpp#L71): Unused private function: 'RedisBlockQuery::getBlockId' [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedPrivateFunction)
1. ![MINOR][MINOR] [redis_block_query.hpp#L81](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/irohad/ametsuchi/impl/redis_block_query.hpp#L81): Unused private function: 'RedisBlockQuery::callbackToLrange' [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedPrivateFunction)
1. ![MINOR][MINOR] [storage_impl.cpp#L36](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/irohad/ametsuchi/impl/storage_impl.cpp#L36): parameter list has 8 parameters, which is greater than the 5 authorized. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3ATooManyParameters)
1. ![MINOR][MINOR] [storage_impl.hpp#L82](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/irohad/ametsuchi/impl/storage_impl.hpp#L82): parameter list has 8 parameters, which is greater than the 5 authorized. [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cxx%3ATooManyParameters)
1. ![MINOR][MINOR] [ametsuchi_test.cpp#L187](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/test/module/irohad/ametsuchi/ametsuchi_test.cpp#L187): Unused variable: account [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedVariable)
1. ![MINOR][MINOR] [ametsuchi_test.cpp#L187](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/test/module/irohad/ametsuchi/ametsuchi_test.cpp#L187): Unused variable: src_account [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedVariable)
1. ![MINOR][MINOR] [ametsuchi_test.cpp#L187](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/test/module/irohad/ametsuchi/ametsuchi_test.cpp#L187): Unused variable: dest_account [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedVariable)
1. ![MINOR][MINOR] [ametsuchi_test.cpp#L187](https://github.com/hyperledger/iroha/blob/502a31fa342937ff1f66694064872ae2742a27c2/test/module/irohad/ametsuchi/ametsuchi_test.cpp#L187): Unused variable: asset [rule](https://sonar.soramitsu.co.jp/coding_rules#rule_key=cppcheck%3AunusedVariable)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: irohad/ametsuchi/impl/block_storage.hpp [hunk 1]
```diff
@@ -79,74 +69,40 @@ namespace iroha {
...
-       */
-      std::string directory() const;
+      boost::optional<std::vector<uint8_t>> get(Identifier id) const;
 
       /**
-       * @return maximal not null key
+       * @return number of identifiers in block storage.
        */
-      Identifier last_id() const;
+      size_t total_keys() const;
```
Comment (luckychess) [inline irohad/ametsuchi/impl/block_storage.hpp]: Why not Identifier type?

File: irohad/ametsuchi/impl/block_storage_nudb.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        } else if (ec) {
+          // some other error occurred
+          return 0;
+        }
+      } while (!found_last);
+
+      return current;
+    }
+
+    std::array<uint8_t, sizeof(uint32_t)> BlockStorage::Impl::serialize_uint32(
```
Comment (luckychess) [inline irohad/ametsuchi/impl/block_storage_nudb.cpp]: Looks like it could be moved to utils.

File: irohad/ametsuchi/impl/block_storage.hpp (hunk 1 above)
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage.hpp]: Identifier = type of key, when `total_keys` returns a number of keys  stored in the database.

File: cmake/Modules/FindNuDB.cmake [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+set(URL https://github.com/warchant/NuDB.git)
+set(VERSION 4535601bb71c578061edd3c3c9098d828ae12b5b) # latest master, 1.0.2
+set_target_description(nudb "Fast KV database" ${URL} ${VERSION})
+
+
+if (NOT nudb_FOUND)
+    externalproject_add(warchant_nudb
+            GIT_REPOSITORY ${URL}
+            GIT_TAG        ${VERSION}
+            GIT_SUBMODULES "doc/docca" # disable submodules.
```
[the remark this answers is not in the document]
Comment (Warchant) [reply cmake/Modules/FindNuDB.cmake]: Not wrong. GIT_SUBMODULES accepts a list of paths to submodules, which we want to download. If empty, all submodules are downloaded. There is no way to disable downloading of all dependencies, therefore I put one "the lightest" dependency here.

File: cmake/Modules/Finded25519.cmake [hunk 4]
```diff
@@ -12,7 +12,7 @@ find_package_handle_standard_args(ed25519 DEFAULT_MSG
     )
 
 set(URL https://github.com/warchant/ed25519.git)
-set(VERSION 56d926c4489aa7b2bb664a48294c5ee2d6e1c283)
+set(VERSION 6b4779f737d72038c716d0ecbc0701411dff083e)
```
[the remark this answers is not in the document]
Comment (Warchant) [reply cmake/Modules/Finded25519.cmake]: I agree, it is bad, but I've already built docker image (hyperledger/iroha-docker-develop:v2) with this change (I don't remember why I changed it in docker), therefore to be consistent with docker, I changed it here.
Sorry.

File: irohad/ametsuchi/impl/block_storage_nudb.cpp (hunk 2 above)
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.cpp]: I agree, but maybe not now. We need separate class or set of utils, which can do serialization and deserialization. For now it is better to leave it here.

File: test/module/irohad/ametsuchi/ametsuchi_fixture.hpp [hunk 5]
```diff
@@ -114,6 +127,9 @@ DROP TABLE IF EXISTS role;
       size_t redisport_ = 6379;
 
       std::string block_store_path = "/tmp/block_store";
+
+     public:
+      static const size_t FIRST_BLOCK = 0;
```
[the remark this answers is not in the document]
Comment (Warchant) [reply test/module/irohad/ametsuchi/ametsuchi_fixture.hpp]: It makes implementation unnecessary complicated. So no, it does not worth it.

File: test/module/irohad/ametsuchi/block_storage_test.cpp [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+          block_store_path,
+          perms::owner_read | perms::owner_write | perms::owner_exe);
+      fs::remove_all(block_store_path);
+    } catch (...) {
+    }
+  }
+
+  void SetUp() override {
+    block = std::vector<uint8_t>(16, 'a');
+    clear();
```
[the remark this answers is not in the document]
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_storage_test.cpp]: I had an issue with primary key duplication, when test is failed or killed with SIGSEKV or another, in these cases TearDown is not called and old database remains in postgres.

File: test/module/vendor/nudb_usage.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+ public:
+
+  void TearDown() override {
+    nudb::erase_file(dat_path);
+    nudb::erase_file(key_path);
+    nudb::erase_file(log_path);
+  }
+};
+
+TEST_F(NuDB, CRUD) {
```
[the remark this answers is not in the document]
Comment (Warchant) [reply test/module/vendor/nudb_usage.cpp]: There is no need in testing of NuDB, it is only usage example test. 
It can be safely commented or removed.

File: irohad/ametsuchi/impl/redis_block_query.cpp [hunk 8]
```diff
@@ -15,28 +15,32 @@
...
+                                     BlockStorage &file_store)
         : block_store_(file_store), client_(client) {}
 
     rxcpp::observable<model::Block> RedisBlockQuery::getBlocks(uint32_t height,
                                                                uint32_t count) {
-      auto last_id = block_store_.last_id();
-      auto to = std::min(last_id, height + count - 1);
-      if (height > to or count == 0) {
+      size_t total = block_store_.total_keys();
+      if (height >= total or count == 0) {
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/redis_block_query.cpp]: total = 10
height = 5
count = 6

then

from=5
to=10

[the remark this answers is not in the document]
Comment (luckychess) [reply irohad/ametsuchi/impl/redis_block_query.cpp]: Ok, but what if total = 10, height = 8, count = 8?

[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/redis_block_query.cpp]: Height+count =16,
To=min(10, 16)=10

[the remark this answers is not in the document]
Comment (luckychess) [reply irohad/ametsuchi/impl/redis_block_query.cpp]: Well but in this case behaviour will be different from the old code.

[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/redis_block_query.cpp]: All tests passed. I still do not understand why. Show please the specific  example.

[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/redis_block_query.cpp]: Ah, I realized that you mean.

File: irohad/ametsuchi/impl/block_storage.hpp [hunk 9]
```diff
@@ -79,74 +69,40 @@ namespace iroha {
...
-       * @param dump_dir - folder of storage
-       * @return - last available identifier
+       * @brief Clear database.
+       * @return true, if success, false otherwise.
        */
-      static nonstd::optional<Identifier> check_consistency(
-          const std::string &dump_dir);
-
-      void dropAll();
+      bool drop_all();
```
Comment (x3medima17) [inline irohad/ametsuchi/impl/block_storage.hpp]: Why don't you use `Result` type, to return error message if `drop_all()` fails?

File: irohad/ametsuchi/impl/block_storage_nudb.cpp [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+    }
+
+    bool BlockStorage::Impl::drop_db() {
+      nudb::error_code ec1, ec2, ec3;
+
+      nudb::erase_file(db_->dat_path(), ec1);
+      nudb::erase_file(db_->log_path(), ec2);
+      nudb::erase_file(db_->key_path(), ec3);
+
+      return !(ec1 || ec2 || ec3);
```
Comment (x3medima17) [inline irohad/ametsuchi/impl/block_storage_nudb.cpp]: Please use `not` instead of `!`, and `or` instead of `||`

File: irohad/ametsuchi/impl/block_storage_nudb.cpp [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+                     current++;
+                   },
+                   ec);
+        if (ec == nudb::error::key_not_found) {
+          return current;
+        } else if (ec) {
+          // some other error occurred
+          return 0;
+        }
+      } while (!found_last);
```
Comment (x3medima17) [inline irohad/ametsuchi/impl/block_storage_nudb.cpp]: Please use `not` instead of `!`

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 12]
```diff
@@ -171,7 +182,7 @@ DROP TABLE IF EXISTS role;
 
       // erase blocks
       log_->info("drop block store");
-      block_store_->dropAll();
+      block_store_->drop_all();
```
Comment (x3medima17) [inline irohad/ametsuchi/impl/storage_impl.cpp]: You don't check if `drop_all` succeeds.

File: test/integration/pipeline/tx_pipeline_integration_test.cpp [hunk 13]
```diff
@@ -23,14 +23,17 @@
 using namespace iroha::model::generators;
 using namespace iroha::model::converters;
 
+#define INTERNAL_PORT 10001
```
Comment (x3medima17) [inline test/integration/pipeline/tx_pipeline_integration_test.cpp]: Why macro instead of constant?

File: test/module/irohad/ametsuchi/block_storage_test.cpp [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+ protected:
+  void clear() {
+    try {
+      namespace fs = boost::filesystem;
+      using perms = boost::filesystem::perms;
+      fs::permissions(
+          block_store_path,
+          perms::owner_read | perms::owner_write | perms::owner_exe);
+      fs::remove_all(block_store_path);
+    } catch (...) {
```
Comment (x3medima17) [inline test/module/irohad/ametsuchi/block_storage_test.cpp]: image

File: test/module/irohad/ametsuchi/block_storage_test.cpp [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+  auto bs = std::move(*s);
+
+  for (uint8_t i = 0x00; i < 0xff; i++) {
+    auto v = std::vector<uint8_t>(16, i);
+    auto total = bs->total_keys();
+    ASSERT_TRUE(bs->add(i, v)) << "can not add block";
+    ASSERT_EQ(total + 1, bs->total_keys());
+
+    auto item = bs->get(i);
+    if (!item) {
```
Comment (x3medima17) [inline test/module/irohad/ametsuchi/block_storage_test.cpp]: Use `not` instead of `!`

File: irohad/ametsuchi/impl/storage_impl.cpp (hunk 12 above)
Comment (Warchant) [reply irohad/ametsuchi/impl/storage_impl.cpp]: Neither previous code.

File: CMakeLists.txt [hunk 16]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.5.1)
+cmake_minimum_required(VERSION 3.5.2)
```
[the remark this answers is not in the document]
Comment (Warchant) [reply CMakeLists.txt]: Not all developers use 3.10.2. 
It is **minimal required**, but actual version can be arbitrary (higher).

File: test/integration/pipeline/tx_pipeline_integration_test.cpp (hunk 13 above)
Comment (Warchant) [reply test/integration/pipeline/tx_pipeline_integration_test.cpp]: There is no difference. The only advantage I see is that in CLion it is highlighted differently.

File: test/module/irohad/ametsuchi/block_storage_test.cpp (hunk 14 above)
Comment (Warchant) [reply test/module/irohad/ametsuchi/block_storage_test.cpp]: Normally you would want to remove test block store regardless of permissions or anything else. Here I do not bother myself which exceptions are thrown here, I don't care. 
This method `clean()` works exactly as I want: removes test dir or does nothing.

File: CMakeLists.txt (hunk 16 above)
[the remark this answers is not in the document]
Comment (lebdron) [reply CMakeLists.txt]: It would be hard to find an error in case some feature from CMake 3.5+ is used, since there will be no error in container build. So either docker CMake version should be 3.5.2, in which case SWIG build should fail due to some missing functions, or version in CMakeLists.txt should be 3.10.2

[the remark this answers is not in the document]
Comment (Warchant) [reply CMakeLists.txt]: If you use any functionality that requires cmake of higher version --  you are welcome to change it. 
I use NuDB, which requires 3.5.2. If SWIG requires cmake 3.7+, someone should conduct tests and make according fix. It is not a part of this task.

**Comment by Warchant** [comment]
WIP

File: irohad/ametsuchi/impl/block_storage_nudb.cpp [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+        Identifier id) const {
+      nudb::error_code ec;
+      boost::optional<std::vector<uint8_t>> ret;
+      auto key = serialize_uint32(id);
+      db_->fetch(key.data(),
+                 [&ret](const void *p, size_t size) {
+                   if (size == 0) {
+                     ret = boost::none;
+                   } else {
+                     const auto *c = static_cast<const char *>(p);
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.cpp]: fixed

File: irohad/ametsuchi/impl/block_storage_nudb.cpp [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+      } else {
+        // dir does not exist, so then create
+        if (not fs::create_directory(path, err)) {
+          log_->error("Cannot create storage dir: {}\n{}", path, err.message());
+          return boost::none;
+        }
+      }
+
+      // paths to NuDB files
+      fs::path dat = fs::path{path} / "iroha.dat";
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.cpp]: wontfix

File: irohad/ametsuchi/impl/block_storage_nudb.hpp [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+
+      /**
+       * Returns total number of keys starting from 0.
+       *
+       * @example in case of 5 keys, then latest id is 4.
+       */
+      size_t total_keys() const override;
+
+      /**
+       * Removes database.
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.hpp]: You should not care how it works internally.

wontfix

File: irohad/ametsuchi/impl/block_storage_nudb.hpp [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+       * Returns path to the storage directory.
+       */
+      const std::string &directory() const override;
+
+      /**
+       * Serializes uint32 as 4-byte big-endian array.
+       * @param t a number
+       * @return array of 4 bytes
+       */
+      static std::array<uint8_t, sizeof(uint32_t)> serialize_uint32(uint32_t t);
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.hpp]: When will we have a separate util class for serialization, it can be extracted there. Currently you should not care how exactly keys are serialized.

File: irohad/ametsuchi/impl/block_storage_nudb.hpp [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+
+     private:
+      /**
+       * Some sort of initialization: it counts number of existent keys in
+       * database starting with 0.
+       * @param db initialized database
+       * @param ec internal error code
+       * @return number of non-empty keys.
+       */
+      static uint32_t count_blocks(nudb::store &db, nudb::error_code &ec);
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.hpp]: It is private method, you should not care how it works internally.

File: irohad/ametsuchi/impl/block_storage_nudb.hpp [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+
+      //< arbitrary number, app-specific
+      static constexpr size_t appid_{0x1337u};
+
+      //< load factor for basket
+      static constexpr float load_factor_{0.5f};
+
+     private:
+      /**
+       * Some sort of initialization: it counts number of existent keys in
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.hpp]: Added to block_storage.hpp

File: irohad/ordering/impl/ordering_gate_transport_grpc.cpp [hunk 23]
```diff
@@ -27,7 +27,7 @@ grpc::Status OrderingGateTransportGrpc::onProposal(
 
   auto transactions = decltype(std::declval<model::Proposal>().transactions)();
   for (const auto &tx : request->transactions()) {
-    transactions.push_back(*factory_.deserialize(tx));
+    transactions.emplace_back(*factory_.deserialize(tx));
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ordering/impl/ordering_gate_transport_grpc.cpp]: Will not cause merge problem, I rebased branch.
wontfix

File: irohad/ordering/impl/ordering_service_impl.cpp [hunk 24]
```diff
@@ -29,7 +32,12 @@ namespace iroha {
           max_size_(max_size),
           delay_milliseconds_(delay_milliseconds),
           transport_(transport),
-          proposal_height(2) {
+
+          // TODO(@warchant):  this is ill-formed, design should be reviewed. It
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ordering/impl/ordering_service_impl.cpp]: This will gone after ordering service is reworked.

File: irohad/ordering/impl/ordering_service_impl.cpp [hunk 25]
```diff
@@ -18,6 +18,9 @@
 #include "ordering/impl/ordering_service_impl.hpp"
 #include "model/peer.hpp"
 
+// TODO(@warchant): move to separate config file (header)
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ordering/impl/ordering_service_impl.cpp]: This will gone after ordering service is reworked.

File: irohad/ametsuchi/impl/block_storage_nudb.cpp [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+                   }
+
+                   // go to the next key
+                   current++;
+                 },
+                 ec);
+        if (ec == nudb::error::key_not_found) {
+          return current;
+        } else if (ec) {
+          // some other error occurred
```
[the remark this answers is not in the document]
Comment (Warchant) [reply irohad/ametsuchi/impl/block_storage_nudb.cpp]: fixed

File: irohad/ametsuchi/impl/block_storage_nudb.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/ametsuchi/impl/block_storage_nudb.hpp]: There's a difference between underlying implementation and what function should do, the last one includes files on which it should operate

File: irohad/ametsuchi/impl/block_storage_nudb.hpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/ametsuchi/impl/block_storage_nudb.hpp]: We have a file for such functions, don't see any problem. Plz make more detailed answer

File: irohad/ametsuchi/impl/block_storage_nudb.hpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/ametsuchi/impl/block_storage_nudb.hpp]: Not really a good argument, private methods shouldn't have bad interfaces just because they're private

File: irohad/ordering/impl/ordering_gate_transport_grpc.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (l4l) [reply irohad/ordering/impl/ordering_gate_transport_grpc.cpp]: I mean other prs, such as #950

File: CMakeLists.txt (hunk 16 above)
[the remark this answers is not in the document]
Comment (neewy) [reply CMakeLists.txt]: Please increase the version so that we won't have a compatibility problem with development dependency (SWIG) @Warchant

File: irohad/ordering/impl/ordering_service_impl.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (neewy) [reply irohad/ordering/impl/ordering_service_impl.cpp]: Please include JIRA ordering service issue id

File: CMakeLists.txt (hunk 16 above)
[the remark this answers is not in the document]
Comment (l4l) [reply CMakeLists.txt]: swig should be unrelated to that pr, why that fixes should be here?

File: test/integration/pipeline/tx_pipeline_integration_test.cpp [hunk 27]
```diff
@@ -79,19 +82,16 @@ class TxPipelineIntegrationTest : public TxPipelineIntegrationTestFixture {
...
-  // TODO 19/12/17 motxx - Rework integration test using shared model (IR-715
-  // comment)
+  // TODO 19/12/17 motxx - Rework integration test using shared model (IR-715)
   // generate test command
-  auto cmd =
-      iroha::model::generators::CommandGenerator().generateAddAssetQuantity(
-          "admin@test",
-          "coin#test",
-          iroha::Amount().createFromString("20.00").value());
+  auto cmd = CommandGenerator().generateAddAssetQuantity(
```
Comment (lebdron) [inline test/integration/pipeline/tx_pipeline_integration_test.cpp]: Unrelated changes.

File: test/integration/pipeline/tx_pipeline_integration_test_fixture.hpp [hunk 28]
```diff
@@ -30,10 +30,14 @@
 #include "module/irohad/ametsuchi/ametsuchi_fixture.hpp"
 #include "integration/pipeline/test_irohad.hpp"
 
+// TODO(@warchant): move to separate configuration file
+#define MAX_WAITING_TIME 11s
```
Comment (lebdron) [inline test/integration/pipeline/tx_pipeline_integration_test_fixture.hpp]: https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros

>Avoid defining macros, especially in headers; prefer inline functions, enums, and const variables.

File: irohad/ordering/impl/ordering_service_impl.cpp [hunk 29]
```diff
@@ -18,6 +18,9 @@
...
+#define FIRST_BLOCK_STARTS_WITH 0
```
Comment (lebdron) [inline irohad/ordering/impl/ordering_service_impl.cpp]: https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros

>Avoid defining macros, especially in headers; prefer inline functions, enums, and const variables.

**Comment by l4l** [comment]
What's the status of this?

**Comment by Warchant** [comment]
Will be splitted onto multiple PRs.

First is https://github.com/hyperledger/iroha/pull/998
